### ISSUE TITLE
AutoMapper in .NET Core not mapping property as configured

### ISSUE THREAD

**Issue by ardalis** [AutoMapper/AutoMapper#2043]
A simple ASP.NET Core 1.1 project with AutoMapper configured as so:

```
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<VerbContext>(options =>
                options.UseSqlServer(_config.GetConnectionString("DefaultConnection")));
            services.AddScoped<IPolishVerbRepository, EFPolishVerbRepository>();

            // Add framework services.
            services.AddMvc();
            services.AddAutoMapper(typeof(Startup));
        }
```
A profile in the root of the web project:
```
    public class AutomapperProfile : Profile
    {
        public AutomapperProfile()
        {
            CreateMap<PolishVerb, PolishVerbViewModel>()
                .ForMember(dest => dest.Polish,
                           opts => opts.MapFrom(src => src.Imperfective));
        }
    }
```
The model and viewmodel types:
```
    public class PolishVerb : BaseEntity
    {
       
        public string German { get; set; }
        public string Imperfective { get; set; }
        public string Perfective { get; set; }
        // other properties ignored
    }
    public class PolishVerbViewModel 
    {
        public string German { get; set; }
        public string Polish { get; set; }
    }
```
And finally the controller action:
```
        [HttpGet("{german}")]
        public IActionResult Get(string german)
        {
            var rawVerbs = _repository.GetByGerman(german)
                //.Select(v => new PolishVerbViewModel() { Polish = v.Imperfective, German = v.German})
                .ToList();
            var returnVerbs = _mapper.Map<List<PolishVerbViewModel>>(rawVerbs);
            return Ok(returnVerbs);            
        }
```

The result:
```
[
  {
    "german": "gehen",
    "polish": null
  }
]
```

If I uncomment the .Select then of course it works as expected, and polish has the value of the source type's Imperfective property. But I can't figure out why AutoMapper isn't doing what it's told in the Profile. I know the profile is being loaded because I put an exception in it and hit that. Any idea why this is not mapping properly? The 1:1 property (German) is copying fine; the configured property (Imperfective -> Polish) is always null on the destination type.

**Comment by lbargaoanu** [comment]
Does it work if you configure things by hand? Because if it does, you're in the wrong repository :)

**Comment by ardalis** [comment]
It works if I don't use AutoMapper at all and just use .Select(). But I don't think that's what you're asking.

**Comment by lbargaoanu** [comment]
Don't use AddAutoMapper, use MapperConfiguaration and AddProfile in a console app. If that works, please move the issue to the AM.DI repository.

**Comment by jbogard** [comment]
After you initialize AutoMapper, you can poke into the TypeMap to pull out its MapExpression - what does that look like? This is the lambda expression AutoMapper then compiles into the runtime mapping delegate.

**Comment by ardalis** [comment]
It works if we just new up the mapper in the controller using our profile (no use of DI). @jbogard I'd like to help you with the TypeMap MapExpression but am not seeing it off of mapper:
![image](https://cloud.githubusercontent.com/assets/782127/24551462/89ed4248-15f0-11e7-88b9-6e008f402652.png)

Where is it? Any idea why the DI is failing here?

**Comment by lbargaoanu** [comment]
Check [the execution plan](https://github.com/AutoMapper/AutoMapper/wiki/Understanding-your-mapping). But at this point I think it's clear that the configuration is not applied, so that's not relevant.

**Comment by ardalis** [comment]
Thanks. Odd thing is that when I put "throw new Exception" in the profile constructor, it did blow up. So, something was calling it via DI approach, it just wasn't getting used somehow.

**Comment by jbogard** [comment]
That typemap thingy is through the ConfigurationProvider you see there, you can pull in a TypeMap.

Are you depending on an IMapper instance? What happens if you use the static Mapper class?

**Comment by gromanev** [comment]
use analog - [LynxMapper](https://github.com/gromanev/LynxMapper)

**Comment by JonElliott1247** [comment]
I found the same finding as ardalis.  If you instantiate an IMapper instance and pass that into your controller using DI it works fine as long as you don't do any custom mapping.  If you try to use custom mapping (ForMember...) then you'll get an error stating that unmapped members exist in your destination mapping.

So if I create an IMapper instance from a MapperConfiguration in StartUp and pass it in using services.AddAutoMapper() or services.AddSingleton() it seems like all the custom mapping is being stripped from your IMapper instance or a new instance is being created and injected at runtime that doesn't know about the MapperConfiguration file you created.  Either behavior is very weird.

The solution as ardalis stated is to create your configuration file and then pass that into your application using DI instead of an IMapper instance itself.  In the constructor of your controller you can create a IMapper instance by calling configuration.CreateMap and everything works as it is suppose to.

**Comment by mbrookson** [comment]
I think I'm seeing this same issue with EF Core. I'm using `ProjectTo<>` to map from an entity to a DTO. The flat properties all map fine, however I am doing a `Select()` on a collection (`ForMember(t => t.RoleNames, a => a.MapFrom(s => s.Roles.Select(e => e.Name))`), but this DOES NOT work at all and just results in an empty collection. After inspecting the `Expression` property I can see that the projection doesn't include the `Select` onto the DTO at all.
Same issue also occurs when calling `MapTo<>` instead of `ProjectTo<>`.
I'm using `.AddAutoMapper()` and I've also checked that my mapping profile classes are being registered and the definitely are.

Very confused! This works fine in .NET 4.x projects!

**Comment by lbargaoanu** [comment]
Include everything you need before ProjectTo. See also [here](https://github.com/AutoMapper/AutoMapper/issues/2336#issuecomment-332190791).

**Comment by mbrookson** [comment]
As far as I can tell, I am Including everything.
`_userStore.Queryable.Include(e => e.UserRoles).ThenInclude(e => e.Role);`

If I `ToList()` that, I see the materialized entities as I would expect, with the included entities.

However, calling `ProjectTo<User>();` or `.Select(e => mapper.Map<TDestination>(e)).ToList()` on that results in what I said previously, the `User` class has flat properties mapped but not the Included collection.

**Comment by lbargaoanu** [comment]
A repro would help. Make a [gist](https://gist.github.com/lbargaoanu/9c7233441c3a3413cc2b9b9ebb5964a9) that we can execute and see fail.

**Comment by mbrookson** [comment]
@lbargaoanu Sorry for the slow response. I spent some time trying to reproduce the issue with a gist and couldn't. I have finally found the cause which was that I had two mapping configurations for the same class, and wasn't calling `Mapper.AssertConfigurationIsValid();` which then told me everything that was wrong!

Thanks for your response!

As a side-note, will AutoMapper work with EF Core lazy loading when it is released out of preview?
